gdnative-async/src/rt/func_state.rs

```diff
@@ -0,0 +1,186 @@
+use gdnative_core::godot_site;
+use gdnative_core::nativescript::init::method::StaticArgs;
+use gdnative_core::nativescript::init::{
+    ClassBuilder, ExportInfo, PropertyUsage, Signal, SignalArgument,
+};
+use gdnative_core::nativescript::method::StaticArgsMethod;
+use gdnative_core::nativescript::user_data::LocalCellData;
+use gdnative_core::nativescript::{
+    Instance, Map, MapMut, NativeClass, NativeClassMethods, RefInstance,
+};
+use gdnative_core::thread_access::{Shared, Unique};
+use gdnative_derive::FromVarargs;
+
+use crate::future::Resume;
+
+pub(crate) struct FuncState {
+    kind: Kind,
+}
+
+enum Kind {
+    Resolved(Variant),
+    Resumable(Resume<Variant>),
+    Pending,
+}
+
+impl NativeClass for FuncState {
+    type Base = Reference;
+    type UserData = LocalCellData<FuncState>;
+
+    fn class_name() -> &'static str {
+        // Sort of just praying that there will be no duplicates of this.
+        "__GDNATIVE_ASYNC_INTERNAL__FuncState"
+    }
+
+    fn register_properties(builder: &ClassBuilder<Self>) {
+        builder.add_signal(Signal {
+            name: "completed",
+            args: &[SignalArgument {
+                name: "value",
+                default: Variant::new(),
+                export_info: ExportInfo::new(VariantType::Nil),
+                usage: PropertyUsage::DEFAULT,
+            }],
+        });
+
+        builder.add_signal(Signal {
+            name: "resumable",
+            args: &[],
+        });
+    }
+}
+
+impl FuncState {
+    pub fn new() -> Instance<Self, Unique> {
+        assert!(
+            super::REGISTRATION.get().is_some(),
+            "async API must be registered before any async methods can be called"
+        );
+
+        Instance::emplace(FuncState {
+            kind: Kind::Pending,
+        })
+    }
+}
+
+pub(super) fn resolve(this: RefInstance<'_, FuncState, Shared>, value: Variant) {
+    this.script()
+        .map_mut(|s| {
+            match s.kind {
+                Kind::Resolved(_) => {
+                    panic!("`resolve` should only be called once for each FuncState")
+                }
+                Kind::Pending => {}
+                Kind::Resumable(_) => {
+                    gdnative_core::log::warn(
+                        Default::default(),
+                        "async function resolved while waiting for a `resume` call",
+                    );
+                }
+            }
+
+            s.kind = Kind::Resolved(value.clone());
+        })
+        .expect("no reentrancy");
+
+    this.base().emit_signal("completed", &[value]);
+}
+
+pub(super) fn make_resumable(this: RefInstance<'_, FuncState, Shared>, resume: Resume<Variant>) {
+    let kind = this
+        .script()
+        .map_mut(|s| std::mem::replace(&mut s.kind, Kind::Resumable(resume)))
+        .expect("no reentrancy");
+
+    match kind {
+        Kind::Resolved(_) => {
+            panic!("`make_resumable` should not be called after resolution")
```

Comment:
Same thought here. Could returning `Result` be helpful? Adding a gdnative_core::log::error() would help make the error more visible.

Reply:
The state can only be set to `Resolved` after the user future resolved and the context is dropped. See `method::Spawner::spawn`. This can only fail due to a bug in the library. There is no way an end user can meaningfully handle this error.